Level 1

yixwellet opened the pull request.

yixwellet committed:
Выполнена начальная настройка проекта:

1. Проинициализирован package.json
2. Установлены все необходимые зависимости
3. Созданы файлы gitignore, eslintrc, editorconfig

yixwellet committed:
прописал в package.json скрипты для dev и start для запуска приложения

yixwellet committed:
Созданы схемы и модели:

1. Создана схема user
2. Создана схема article

yixwellet committed:
скопироаны файлы ошибок и мидлвэров

yixwellet committed:
Созданы роуты и контроллеры.

1. получение информации о пользователе
2. получение списка сохраненных статей
3. создание статьи
4. удаление статьи

yixwellet committed:
Отредактирован README

yixwellet committed:
editorconfig удален из gitignore

yixwellet committed:
Update README.md

yixwellet committed:
Merge branch 'level-1' of https://github.com/yixellet/news-explorer-api into level-1

Boortcore commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id _;
- [ ] <!--3.64--> Нет ошибок линтера.

Проект не требует никаких переопределений правил, кроме  "no-underscore-dangle", потому как MongoDB использует свойство с нижним подчёркиванием в своём названии - _id. Все остальные переопределеня правил излишни. Их использование приводит к написанию кода, который перестаёт соответствовать правилам линтинга.

- [x] <!--3.64--> В разделе scripts файла package.json:
  - [x] есть команда npm run start, которая запускает сервер на localhost:3000;
  - [x] и npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--12.57--> Когда все зависимости установлены, приложение запускается командой npm run dev без ошибок.
- [ ] <!--25.45--> Описанные роуты работают корректно:
  - [x] запрос на GET /users/me возвращает информацию о пользователе (email и имя);

у запроса нет параметра id. Эта валидация не работает, так как используется метод unknown, который допускает отсутствие описанных параметров.
![image](https://user-images.githubusercontent.com/17438227/94527641-f6199900-023f-11eb-8f5a-a2ee4c9c7f38.png)


  - [x] GET /articles — все сохранённые статьи;
  - [ ] POST /articles создаёт статью с переданными в теле данными;

Свойство date по условию задания должна быть строкой. 

  - [x] DELETE /articles/articleId удаляет статью по _id _; 
  - [x] POST /signup создаёт пользователя с переданными в теле данными;
  - [x] POST /signin возвращает JWT, если в теле запроса переданы правильные почта и пароль.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты статей описаны в отдельных файлах.
- [x] <!--3.64--> Ошибки API обрабатываются:
  - [x] если в запросе что-то не так, сервер возвращает ответ с ошибкой и её статусом;
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [ ] <!--3.64--> В production-режиме адрес базы данных берётся из process.env.

Адрес прописан статически и не зависит от режима запуска приложения

- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в зашифрованном виде;
  - [x] API не возвращает хеш пароля клиенту.
- [x] <!--3.64--> Данные валидируются перед добавлением в базу.
- [x] <!--3.64--> Пользователь не может удалить сохранённую карточку из профиля другого пользователя.
- [x] <!--3.64--> К серверу можно обратиться и по http, и по https, обращаясь к домену, указанному в README.md.
- [x] <!--3.64--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--1.5--> Асинхронные операции реализованы промисами или async/await.
- [ ] <!--1.5--> Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры, проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.

Данная валидация не работает. Ожидается параметр articleId, но осуществляется валидация cardId
![image](https://user-images.githubusercontent.com/17438227/94527807-38db7100-0240-11eb-8a0d-66648038d066.png)

- [x] <!--1.5--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [ ] <!--1.5--> Ошибки обрабатываются централизованным обработчиком.

![image](https://user-images.githubusercontent.com/17438227/94527911-5f99a780-0240-11eb-85b6-e26a1f67747b.png)

- [ ] <!--1.5--> Централизованный обработчик ошибок описан в отдельном модуле.
- [ ] <!--1.5--> В случае ошибки API возвращает статус ответа, который соответствует типу ошибки.

Попытка создать статью с невалидным свойством link или image должна возвращать статус 400, сейчас возвращается статус 500
Аналогичная ситуация при удалении карточки, которой не существует в базе.
![image](https://user-images.githubusercontent.com/17438227/94527108-2c0a4d80-023f-11eb-9619-1fd9c0e0064e.png)

- [x] <!--1.5--> К серверу можно обратиться по https по адресу домена, указанному в README.md.
- [ ] <!--1.5--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone
- [ ] <!--1.5--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.

### Рекомендации
- [x] <!--0.83--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [ ] <!--0.83--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [ ] <!--0.83--> Конфигурация и константы в отдельных файлах:
  - [ ] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [ ] секретный ключ для JWT в режиме разработки вынесен в отдельный конфигурационный файл;
  - [ ] константы приложения — сообщения ответов и ошибок — вынесены в отдельный файл с константами.
- [ ] <!--0.83--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [ ] <!--0.83--> rate limiter сконфигурирован в отдельном файле и импортируется в app.js
- [ ] <!--0.83--> API размещено на отдельном поддомене, например: api.news-explorer.tk.

## Количество баллов: 54

## Доп. Комментарии:
- Нет необходимости записывать id пользователя в куки. Для этого есть мидлвара auth, которая записывает информацию о пользователя в свойство user объекта запроса
- ServerError - это ошибка по умолчанию, она генерируется централизованным обработчиком ошибок. Создавать класс ошибки и пробрасывать этот тип ошибки из контроллера не нужно. 
Эти два блока catch равносильны
![image](https://user-images.githubusercontent.com/17438227/94527497-b8b50b80-023f-11eb-8dbb-fb5aecd97e3d.png)
- В контроллере удаления статьи запрос на получение статьи осуществляется дважды: сначала методом Article.findById, а затем Article.findByIdAndRemove. Вместо последнего имеет смысл использовать Article.deleteOne(article), где article - это ранее возвращенный методом Article.findById экземпляр статьи.

yixwellet committed:
Код откорректирован в соответствии с замечаниями.

1. Из настроек Линтера удалены недопустимые правила
2. Celebrate-валидация вынесена в отдельный файл
3. Адрес сервера БД, порт запуска приложения и JWT-ключ вынесены в отдельный файл
4. Исправлена ошибка валидации ID статьи
5. Общий обработчик ошибок вынесен в отдельный файл
6. Исправлены ошибка с некорректными статусами ответов
7. Добавлен модуль Helmet
8. Сообщения об ошибках вынесены в отдельный файл

yixwellet committed:
Отредактирован README

Boortcore commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
  - [x] в проекте есть package.json;
  - [x] .editorconfig;
  - [x] .eslintrc, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
  - [x] .gitignore.
  - [x] в .eslintrc добавлено исключение для _id _;
- [x] <!--3.64--> Нет ошибок линтера.
- [x] <!--3.64--> В разделе scripts файла package.json:
  - [x] есть команда npm run start, которая запускает сервер на localhost:3000;
  - [x] и npm run dev, которая запускает сервер на localhost:3000 с хот релоудом.
- [x] <!--12.57--> Когда все зависимости установлены, приложение запускается командой npm run dev без ошибок.
- [x] <!--25.45--> Описанные роуты работают корректно:
  - [x] запрос на GET /users/me возвращает информацию о пользователе (email и имя)
  - [x] GET /articles — все сохранённые статьи;
  - [x] POST /articles создаёт статью с переданными в теле данными;
  - [x] DELETE /articles/articleId удаляет статью по _id _; 
  - [x] POST /signup создаёт пользователя с переданными в теле данными;
  - [x] POST /signin возвращает JWT, если в теле запроса переданы правильные почта и пароль.
- [x] <!--3.64--> Все роуты, кроме /signin и /signup, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты статей описаны в отдельных файлах.
- [x] <!--3.64--> Ошибки API обрабатываются:
  - [x] если в запросе что-то не так, сервер возвращает ответ с ошибкой и её статусом;
  - [x] асинхронные обработчики завершены блоком catch;
  - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> В production-режиме адрес базы данных берётся из process.env.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
  - [x] пароль хранится в зашифрованном виде;
  - [x] API не возвращает хеш пароля клиенту.
- [x] <!--3.64--> Данные валидируются перед добавлением в базу.
- [x] <!--3.64--> Пользователь не может удалить сохранённую карточку из профиля другого пользователя.
- [x] <!--3.64--> К серверу можно обратиться и по http, и по https, обращаясь к домену, указанному в README.md.
- [x] <!--3.64--> Правильно реализовано хранение секретного ключа для создания JWT:
  - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
  - [x] в режиме разработки (при process.env.NODE_ENV !== 'production') код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--1.5--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--1.5--> Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры, проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
- [x] <!--1.5--> Настроено логгирование:
  - [x] запросы и ответы записываются в файл request.log;
  - [x] ошибки записываются в файл error.log;
  - [x] файлы логов не добавляются в репозиторий.
- [ ] <!--1.5--> Ошибки обрабатываются централизованным обработчиком.

![image](https://user-images.githubusercontent.com/17438227/94903047-dc778c00-04a1-11eb-93d3-807bdf23470e.png)

- [x] <!--1.5--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--1.5--> В случае ошибки API возвращает статус ответа, который соответствует типу ошибки.
- [x] <!--1.5--> К серверу можно обратиться по https по адресу домена, указанному в README.md.
- [ ] <!--1.5--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone
- [ ] <!--1.5--> Все роуты подключены в файле index.js, который находится в папке routes. Оттуда единый роут подключается в файле app.js.
![image](https://user-images.githubusercontent.com/17438227/94903168-0fba1b00-04a2-11eb-8d3a-d10b29dbbcc8.png)


### Рекомендации
- [x] <!--0.83--> Для ошибок API созданы классы, расширяющие конструктор Error.
- [x] <!--0.83--> Используется модуль Helmet для установки заголовков, связанных с безопасностью.
- [x] <!--0.83--> Конфигурация и константы в отдельных файлах:
  - [x] адрес Mongo-сервера в режиме разработки вынесен в отдельный конфигурационный файл;
  - [x] секретный ключ для JWT в режиме разработки вынесен в отдельный конфигурационный файл;
  - [x] константы приложения — сообщения ответов и ошибок — вынесены в отдельный файл с константами.
- [ ] <!--0.83--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено.
- [ ] <!--0.83--> rate limiter сконфигурирован в отдельном файле и импортируется в app.js
- [x] <!--0.83--> API размещено на отдельном поддомене, например: api.news-explorer.tk.

## Количество баллов: 54

yixwellet committed:
Update README.md

yixwellet committed:
Update README.md

yixwellet committed:
Update README.md

yixwellet committed:
Some little changes

yixwellet committed:
Merge branch 'level-1' of https://github.com/yixellet/news-explorer-api into level-1
commit

yixwellet committed:
config changes

yixwellet committed:
config changes

yixwellet committed:
config changes

yixwellet committed:
config changes

yixwellet committed:
config changes

yixwellet committed:
config changes

yixwellet committed:
config changes

yixwellet committed:
config changes

yixwellet committed:
config changes

yixwellet committed:
config changes

yixwellet committed:
config changes

yixwellet committed:
config changes

yixwellet committed:
Update README.md

yixwellet committed:
config changes

yixwellet committed:
Merge branch 'level-1' of https://github.com/yixellet/news-explorer-api into level-1
x
